queryable-spec.ts

```diff
@@ -0,0 +1,421 @@
+   */
+  metadata(opts?: FilterableResultMetadataOptions): Promise<FilterableResultMetadata>;
+  
+  /**
+   * Asynchronously returns a boolean indicating if the requested expression is
+   * supported by the FilterableSource. If it returns true, quads() and 
+   * metadata() MAY produce a valid result. If it returns false, quads() MUST
+   * return a stream emitting an error, and metadata() MUST reject.
+   */
+  isSupported(): Promise<boolean>;
+};
+
+/* 
+ * A FilterableSource is an object that produces a FilterableSourceResult that
+ * can emit quads. The emitted quads can be directly contained in this 
+ * FilterableSourceo bject, or they can be generated on the fly.
+ * 
+ * FilterableSource is not necessarily an extension of the RDF/JS Source 
+ * interface, but implementers MAY decide to implement both at the same time.
+ * 
+ * matchExpression() Returns a FilterableSourceResult that contains a quad 
+ * stream that processes all quads matching the quad pattern and the expression.
+ * 
+ * When a Term parameter is defined, and is a NamedNode, Literal or BlankNode, 
+ * it must match each produced quad, according to the Quad.equals semantics. 
+ * When a Term parameter is a Variable, or it is undefined, it acts as a 
+ * wildcard, and can match with any Term.
+ * 
+ * NOTES:
+ * - When matching with graph set to undefined or null it MUST match all the
+ *   graphs (sometimes called the union graph). To match only the default graph
+ *   set graph to a DefaultGraph.
+ * - When an Expression parameter is defined, the complete quad stream is
+ *   filtered according to this expression. When it is undefined, no filter is
+ *   applied.
+ * 
+ * If parameters of type Variable with an equal variable name are in place,
+ * then the corresponding quad components in the resulting quad stream MUST be
+ * equal.
+ * Expression's MAY contain Variable Term's. If their variable names are equal
+ * to Variable's in the given quad pattern, then the Expression MUST be 
+ * instantiated for each variable's binding in the resulting quad stream when
+ * applying the Expression filter.
+ */
+interface FilterableSource {
+  matchExpression(
+    subject?: RDF.Term,
+    predicate?: RDF.Term,
+    obj?: RDF.Term,
+    graph?: RDF.Term,
+    expression?: Expression,
+    opts?: { 
+      limit?: number; 
+      offset?: number; 
+    },
+  ): FilterableResult;
+};
+
+
+/******************************************************************************
+                              QUERYABLE SOURCE
+ *****************************************************************************/
+
+/*
+ * Map-like representation of Bindings as using plain objects could lead
+ * to collisions between variable names and object properties. Support for
+ * immutability is required (but implementations are free to be mutable) which
+ * determines the return value of the set() and delete() methods to be an 
+ * instance of Bindings (potentially a different one).
+ */ 
```

Comment:
> It would also help if it defined Symbol.iterator to iterate entries, so that it could trivially be used to construct a Map.

+1

> Why does it need to ever be mutable? Can we ditch the mutation methods?

@rubensworks ? I'm open to this.

Reply:
If we keep them in the spec, then I'm in favor of the current approach as I find it more familiar (i.e. it's closer to how the native `Map` works).

Reply:
I think this familiary might actually cause problems in practise, as developers could invoke `bindings.set(...)`, expecting that the original `bindings` object will be mutated, while it in fact returned a new instance and kept the old one unchanged.
Explicitly moving `.set` (or perhaps better: `.with`) to the factory would make this difference in semantics explicit.

Reply:
I've added both changes.  Kept the same name for the `set()` method as we also have `mergeWith()` method in which the `-With` part indicates the use of a custom conflict resolution function.